community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java

```diff
@@ -120,147 +140,90 @@ public PropertyItem get()
-                }
-                else
+                do
                 {
-                    buffer.putLong( cursor.getLong() );
-                    remainingBlocksToRead--;
+                    cursor.setOffset( offset );
+                    for ( int i = 0; remainingBlocksToRead-- > 0; i++ )
+                    {
+                        valuesAfterHeader[i] = cursor.getLong();
+                    }
                 }
-                break;
+                while ( cursor.shouldRetry() );
             }
-
-            case FLOAT:
+            catch ( IOException e )
             {
-                buffer.putInt( (int) header );
-                break;
+                throw new UnderlyingStorageException( e );
             }
+        }
+    }
 
-            case DOUBLE:
-            {
-                buffer.putLong( cursor.getLong() );
-                remainingBlocksToRead--;
-                break;
-            }
-
-            case STRING:
-            {
-                int storeOffset = cursor.getOffset();
-                cursor.close();
-
-                if ( stringRecordCursor == null )
-                {
-                    stringRecordCursor = stringStore.newDynamicRecordCursor();
-                }
-
-                try ( GenericCursor<DynamicRecord> stringRecords = stringStore.getRecordsCursor( header, true,
-                        stringRecordCursor ) )
-                {
-                    while ( stringRecords.next() )
-                    {
-                        DynamicRecord dynamicRecord = stringRecords.get();
-
-                        buffer.put( dynamicRecord.getData(), 0, dynamicRecord.getData().length );
-                    }
-                }
+    private void readFromStore( AbstractDynamicStore store ) throws IOException
+    {
+        int storeOffset = cursor.getOffset();
+        cursor.close();
+        cursor = null;
 
-                cursor = propertyStore.newReadCursor( currentRecordId );
-                cursor.setOffset( storeOffset );
-                break;
-            }
+        if ( recordCursor == null )
+        {
+            recordCursor = store.newDynamicRecordCursor();
+        }
 
-            case ARRAY:
+        buffer.clear();
+        long startBlockId = (header & 0xFFFFFFFFF0000000L) >>> 28;
+        try ( GenericCursor<DynamicRecord> records = store.getRecordsCursor( startBlockId, true, recordCursor ) )
+        {
+            while ( records.next() )
             {
-                int storeOffset = cursor.getOffset();
-                cursor.close();
-
-                if ( arrayRecordCursor == null )
-                {
-                    arrayRecordCursor = arrayStore.newDynamicRecordCursor();
-                }
-
-                try ( GenericCursor<DynamicRecord> arrayRecords = arrayStore.getRecordsCursor( header, true,
-                        arrayRecordCursor ) )
+                DynamicRecord dynamicRecord = records.get();
+                while ( true )
```

Comment:
Because we slowly increase the size of the buffer by 2, probably it is better to create it big enough in one go.